Adding support for passing service ID to NotificationSystem on deletes

cgtz opened the pull request:
- Adding a service ID argument to the Router.deleteBlob() API
- Extracting the service ID from headers in AmbryBlobStorageService
- Adding a service ID argument to the NotificationSystem.onBlobDeleted()

*Reviewers:* @pnarayanan, @xiahome 
*Time to review:* 15 min

pnarayanan commented:
@cgtz The new fields for PUTS - blobType, blobSize, ContainerId - were you planning to add those in another patch?

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/router/Router.java:
```diff
@@ -73,17 +73,18 @@
-   * @param callback The {@link Callback} which will be invoked on the completion of a request.
-   * @return A future that would contain information about whether the deletion succeeded or not, eventually.
+   * @param serviceId The service ID of the service deleting the blob. This can be null if unknown.
+   *@param callback The {@link Callback} which will be invoked on the completion of a request.  @return A future that would contain information about whether the deletion succeeded or not, eventually.
```
fix formatting.

cgtz replied:
fixed

pnarayanan commented on ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java:
```diff
@@ -445,15 +449,20 @@ public void onBlobDeleted(String blobId) {
+      Assert.assertTrue("All blobs in server are deleted", blobsThatAreDeleted.keySet().containsAll(blobsToBeDeleted));
+      Assert.assertTrue("Only blobs in server are deleted", blobsToBeDeleted.containsAll(blobsThatAreDeleted.keySet()));
+      for (Map.Entry<String, String> blobIdAndServiceId : blobsThatAreDeleted.entrySet()) {
+        String expectedServiceId =
```
could you also add that this is getting tested to the javadoc?

cgtz replied:
added

pnarayanan commented on ambry-tools/src/main/java/com.github.ambry/tools/perf/rest/PerfRouter.java:
```diff
@@ -156,21 +156,22 @@ public void onCompletion(Long result, Exception exception) {
-   * @param callback the {@link Callback} to invoke on operation completion.
-   * @return a {@link FutureResult} that will eventually contain the result of the operation.
+   * @param serviceId (ignored).
+   *@param callback the {@link Callback} to invoke on operation completion.  @return a {@link FutureResult} that will eventually contain the result of the operation.
```
something wrong with your formatting template?

cgtz replied:
fixed, for some reason my formatter does not fix comments that go over the line length limit.

cgtz commented:
@pnarayanan The changes for PUT audit logging do not need any changes to open source, as all of the things needed (except for containerID) are in `BlobProperties`.

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java:
```diff
@@ -260,15 +261,16 @@ private void initiateBackgroundDeletes(List<StoreKey> idsToDelete) {
-          currentBackgroundOperationsCount.decrementAndGet();
-        }
-      });
+      backgroundDeleter.deleteBlob(key.getID(), routerConfig.routerBackgroundDeleteServiceId, new FutureResult<Void>(),
```
Would this be a little bit general? Maybe include the service this chunk blob belongs to?

cgtz replied:
Appended the requestor service id for background deletes

xiahome commented on ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java:
```diff
@@ -111,6 +111,13 @@
   /**
+   * The service ID used for background delete operations.
+   */
+  @Config("router.background.delete.service.id")
```
Why this is a config rather than a constant defined in router?

cgtz replied:
changed to a constant

cgtz commented:
@xiahome @pnarayanan Just updated the PR. I added the additional info to the BG delete service ID, and passed in info about blob type on puts. I also added notification on data chunk creation.

xiahome commented on ambry-api/src/main/java/com.github.ambry/notification/CreatedBlobType.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Denotes the type of blob created. i.e. whether this is a simple blob, a metadata (composite) blob, or a data chunk
+ */
+public enum CreatedBlobType {
```
Minor: `BlobType` should be fine, and should not be confused with `ContentType`.

cgtz replied:
changed it to `BlobType` as suggested, since the msgformat `BlobType` isn't used in `PutManager`, where `onBlobCreated` is called.

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/BackgroundDeleteRequest.java:
```diff
@@ -0,0 +1,51 @@
+ * This class contains the request parameters for a background delete operation.
+ */
+class BackgroundDeleteRequest {
+  static final String SERVICE_ID_PREFIX = "ambry-bg-delete-";
```
expand to `ambry-background-delete-` if we are not that conservative on string length?

cgtz replied:
will do

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -245,9 +247,14 @@ void onComplete(PutOperation op) {
+      routerCallback.scheduleDeletes(op.getSuccessfullyPutChunkIdsIfComposite(), op.getServiceId());
     } else {
-      notificationSystem.onBlobCreated(op.getBlobIdString(), op.getBlobProperties(), op.getUserMetadata());
+      CreatedBlobType blobType = op.getNumDataChunks() == 1 ? CreatedBlobType.Simple : CreatedBlobType.Composite;
```
Is this the "recommended" way of checking the blob type?

cgtz replied:
This is how it is done in elsewhere in PutManager/PutOperation. See `updateChunkingAndSizeMetricsOnSuccessfulPut()`

pnarayanan commented on ambry-api/src/main/java/com.github.ambry/notification/BlobType.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Denotes the type of blob created. i.e. whether this is a simple blob, a metadata (composite) blob, or a data chunk
+ */
+public enum BlobType {
```
There is a `BlobType` in MessageFormat api as well which is similar but is used to distinguish between blob content types - whether a blob contains data or metadata. Simple blobs as well as data chunk blobs are the same as far as MessageFormat is concerned.

This is similar, but I think here we are concerned more about what the blob id represents, which is notification system specific and for that does it make more sense to associate this type with the blob id? `BlobIdType` perhaps?

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java:
```diff
@@ -288,7 +291,11 @@ public void onCompletion(GetBlobResultInternal result, Exception exception) {
-          initiateBackgroundDeletes(result.storeKeys);
+          List<BackgroundDeleteRequest> deleteRequests = new ArrayList<>(result.storeKeys.size());
+          for (StoreKey storeKey : result.storeKeys) {
+            deleteRequests.add(new BackgroundDeleteRequest(storeKey, serviceId));
```
so we have decided not to prefix with "bg-delete-"?

cgtz replied:
The prefix is still added on in the BackgroundDeleteRequest constructor. So, if "blah" was provided as the serviceIdSuffix argument, the service ID used for the delete will be "ambry-background-delete-blah"

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -245,9 +247,14 @@ void onComplete(PutOperation op) {
       blobId = null;
       routerMetrics.onPutBlobError(e);
-      routerCallback.scheduleDeletes(op.getSuccessfullyPutChunkIdsIfComposite());
+      routerCallback.scheduleDeletes(op.getSuccessfullyPutChunkIdsIfComposite(), op.getServiceId());
```
Not sure about this. The calling service id is not the one that issued a delete. I think we should at least have an associated prefix, preferably different from that for the deletes triggered by a deleteBlob() call.

cgtz replied:
As mentioned above, this is actually just the service ID suffix, which will be appended to "ambry-background-delete-". I will add something to differentiate it from regular deletes, though. Maybe "ambry-background-delete-failed-put-serviceid"

cgtz commented:
@pnarayanan @xiahome 
Updated the PR. Changed `BlobType` to `NotificationBlobType`, to create a distinction between this "type" and the message format one. Blob size for data chunks is now logged correctly too.

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -1135,6 +1152,34 @@ private void finalizeMetadataChunk() {
+     * @return the list of pairs of successfully put chunk ids and chunk and their {@link BlobProperties} if this is a
+     *         composite object, empty list otherwise.
+     */
+    List<Pair<StoreKey, BlobProperties>> getSuccessfullyPutChunkIdsAndPropertiesIfComposite() {
```
This looks correct, but seems a bit more involved than I imagined. Also, I just realized that if we only initiate when operation completes, then we do not send notifications for successfully put chunk ids for a failed put. Let us discuss and resolve this today.

cgtz replied:
Okay, I have an alternate scheme that we can discuss once you're in office.

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -197,6 +202,22 @@ boolean isOperationComplete() {
+   */
+  void maybeNotifyForBlobCreation() {
+    if (isOperationComplete()) {
+      metadataPutChunk.maybeNotifyForFirstChunkCreation();
```
couldn't this be:
```java
if (getNumDataChunks() > 1) {
  notifyForFirstChunkCreation();
}
```

cgtz replied:
Refactored this a little bit, hopefully it's a bit cleaner now. I can't call `getNumDataChunks()` if chunk filling did not complete successfully, so it's not quite this simple.

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -1088,6 +1118,35 @@ void addChunkId(BlobId chunkBlobId, int chunkIndex) {
+      if (firstChunkIdAndProperties != null) {
+        String chunkId = firstChunkIdAndProperties.getFirst().getID();
+        BlobProperties chunkProperties = firstChunkIdAndProperties.getSecond();
+        if (!chunkId.equals(getBlobIdString())) {
```
And this could simply omit this check - as with the previous change this method will be called if and only if the first chunk notification has to be done.

cgtz replied:
Removed this check

cgtz commented:
@pnarayanan @xiahome formatted and built/tested on osx/linux

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -137,25 +142,26 @@
    * @param routerCallback The {@link RouterCallback} to use for callbacks to the router.
    * @param time the Time instance to use.
-   * @throws RouterException if there is an error in constructing the PutOperation with the given parameters.
+   * @param blobProperties the BlobProperties associated with the put operation.        @throws RouterException if there is an error in constructing the PutOperation with the given parameters.
```
formatting.

cgtz replied:
fixed

pnarayanan commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -1088,6 +1127,33 @@ void addChunkId(BlobId chunkBlobId, int chunkIndex) {
+     * blob is composite. If no first chunk was put successfully, this will do nothing.
+     */
+    void notifyForFirstChunkCreation() {
+      if (firstChunkIdAndProperties != null) {
```
Maybe omit this check since this method will now only be called if it is known for sure that there is a first chunk that needs notification (which is also why it is no longer prefixed with "maybe").

cgtz replied:
made change

cgtz commented:
@pnarayanan Made suggested changes, autoformatted, built/tested on osx/linux
